src/connection.rs

```diff
@@ -34,100 +44,244 @@ impl<C: Connector> EppConnection<C> {
+        self.greeting = RequestFuture { conn: self }.await?;
         Ok(())
     }
 
-    /// Receives response from the socket and converts it into an EPP XML string
-    async fn get_epp_response(&mut self) -> Result<String, Error> {
-        let mut buf = [0u8; 4];
-        timeout(self.timeout, self.stream.read_exact(&mut buf)).await?;
-
-        let buf_size: usize = u32::from_be_bytes(buf).try_into()?;
-
-        let message_size = buf_size - 4;
-        debug!("{}: Response buffer size: {}", self.registry, message_size);
-
-        let mut buf = vec![0; message_size];
-        let mut read_size: usize = 0;
-
-        loop {
-            let read = timeout(self.timeout, self.stream.read(&mut buf[read_size..])).await?;
-            debug!("{}: Read: {} bytes", self.registry, read);
-
-            read_size += read;
-            debug!("{}: Total read: {} bytes", self.registry, read_size);
-
-            if read == 0 {
-                return Err(io::Error::new(
-                    io::ErrorKind::UnexpectedEof,
-                    format!("{}: unexpected eof", self.registry),
-                )
-                .into());
-            } else if read_size >= message_size {
-                break;
-            }
-        }
-
-        self.ready = true;
-        Ok(String::from_utf8(buf)?)
-    }
-
     pub(crate) async fn reconnect(&mut self) -> Result<(), Error> {
         debug!("{}: reconnecting", self.registry);
-        self.ready = false;
+        let _ = self.current.take();
+        let _ = self.next.take();
         self.stream = self.connector.connect(self.timeout).await?;
-        self.greeting = self.get_epp_response().await?;
-        self.ready = true;
+        self.read_greeting().await?;
         Ok(())
     }
 
     /// Sends an EPP XML request to the registry and return the response
     /// receieved to the request
-    pub(crate) async fn transact(&mut self, content: &str) -> Result<String, Error> {
-        if !self.ready {
-            debug!("{}: connection not ready", self.registry);
-            self.reconnect().await?;
-        }
-
-        debug!("{}: request: {}", self.registry, content);
-        self.send_epp_request(content).await?;
+    pub(crate) fn transact<'a>(&'a mut self, command: &str) -> Result<RequestFuture<'a, C>, Error> {
+        debug!("{}: request: {}", self.registry, command);
+        let new = RequestState::new(command)?;
 
-        let response = self.get_epp_response().await?;
-        debug!("{}: response: {}", self.registry, response);
+        // If we have a request currently in flight, finish that first
+        // If another flight was queued up behind the one in flight, just replace it
+        match self.current.is_some() {
+            true => {
+                debug!(
+                    "{}: Queueing up request in order to finish in-flight request",
+                    self.registry
+                );
+                self.next = Some(new);
```

Comment:
Do we need a Vec of queued requests here? With this solution, is the likelihood that our problem will occur 0? Or just reduced?

Reply:
I had a `Vec` at first, but it occurred to me that we don't need it. We only care about finishing any requests that are currently in-flight (because they mess up our state). However, any requests that haven't been started yet (and have already been dropped, by the fact that a caller was able to queue up another one through a mutable reference to the `EppConnection`) can safely be overwritten with our new request.